### ISSUE TITLE
Is behat dead?

### ISSUE THREAD

**Issue by uuf6429** [Behat/Behat#1465]
Issues and pull requests are piling up, core devs seem to be absent or at least focused on other projects (worryingly, competing projects at times).

Looking at the release history, it seems there's one release per year. Somehow, the last 2 releases missed a 5-liner PR I wrote last year (among other simplistic PRs).

I get it everyone is busy, no one wants to be pushed around because no one is paid for it - but contributors are also not paid, in fact we are under a lot of pressure to explain continued use of a product that is poorly supported in the face of fancier alternatives that the cool kids are using nowadays.

If there aren't enough people to handle this, it would make sense to find more _before_ everyone has effectively left... If you don't see a future in the project, deprecate/archive it so everyone can simply jump ship.

**Comment by pierresh** [comment]
I also love using Behat and I am ok to help if needed.

**Comment by swaibar** [comment]
> The only alternative for Story BDD (using Gherkin) in PHP I'm aware of is Codeception

Maybe only an alternative if it didn't use `behat/beaht` for its parser...

**Comment by carlos-granados** [comment]
Last commit was 9 months ago... 😞

**Comment by carlos-granados** [comment]
@stof Do you have any information about the status of this project? It seems pretty abandoned by @everzet which would be a shame as Behat is such an amazing and useful tool. Would it be necessary to look for new maintainers?

**Comment by swaibar** [comment]
@ciaranmcnulty is the current owner I thought

**Comment by carlos-granados** [comment]
@ciaranmcnulty can you confirm?

**Comment by carlos-granados** [comment]
I saw this related issue https://github.com/Behat/Behat/issues/1296 @pamil @acoulton any comments or information about this?

**Comment by carlos-granados** [comment]
I don't really have a deep knowledge of Behat but willing to help to keep this project alive. But would need some kind of response from the maintainers, so can someone please let us know anything?

**Comment by acoulton** [comment]
@carlos-granados sorry, I missed the original ping on this. We are still actively using behat on a daily basis, but since those posts in 2021 we've not really had time to contribute back. Behat is currently working as-expected for us on PHP 8.3, so updating/changing it is less of a priority for us at the moment. 

That could obviously change if we run into issues upgrading to PHP 8.4 (which we'll likely be doing about a year from now : our business cycle means we usually upgrade the autumn after a release, unless something comes up sooner). I see there are a few 8.4-related issues.

I'd still be open to joining a team to occasionally review issues & PRs, but would probably be unlikely to be contributing code myself in the short/medium term unless something breaks for us.

tl;dr (in our specific usecases, YMMV), behat isn't dead so much as it's stable & doesn't need much attention at the moment.

**Comment by carlos-granados** [comment]
Thanks @acoulton, that's helpful. I think Behat needs at least some low-level maintenance to look into issues and PRs and keep it updated for new PHP versions. Again, willing to help here

**Comment by acoulton** [comment]
@carlos-granados I agree - I guess just being realistic and saying that although I'm happy to help in principle, I won't really be able to do much maintenance myself until it bites us. 

But I'd still be happy to discuss becoming a notional maintainer to help review & merge PRs and keep things ticking over - I've been around a while and am not planning on going anywhere soon.

**Comment by uuf6429** [comment]
The problem I see is that if the community decides to make a fork, (among other problems), some (currently unresponsive) owner in the future could always come back to work on the official repo, effectively ruining the community effort.

This is why I, and many others, are asking for a clarification on where this project is at. It's not a matter of contributing - there's enough people willing to contribute on top of the many open pull requests.

**Comment by everzet** [comment]
@carlos-granados @acoulton thank you both so much for offering help. Whatever amount you can offer, it is indeed very welcome.

As you've found in #1296, due to diverging professional trajectory, I've moved away from PHP ecosystem and, as a consequence, Behat. Before I signed off, I've found a new maintainer from the company that heavily relies on it. But it seems similar situation happened with them, albeit without a new maintainer being nominated. So we should correct that.

From what I've heard, Behat is in a pretty stable state feature-wise and doesn't need active development. Instead, as you rightfully pointing out, it needs somebody to look through issues, occasionally bump dependencies, check compatibility and release a new minor version based on PRs people submit. I'd be happy to onboard whoever is willing to commit their time for that.

I'm also willing to dedicate a little bit of my time to help onboard that new maintainer. Happy to provide access, permissions, guidance and insight during that time, to the best of my memory.

How should we proceed?

**Comment by everzet** [comment]
@uuf6429 I am grateful for you opening this issue as it might lead to finding a new active maintainer.

Also thank you for opening multiple PRs, most of which seem to be genuine improvements to Behat. I'll have a closer look at them in coming days.

Aside from that, having read the actual text of your original post and the follow-up above, I'm really not a fan of your language and the attitude it points to:

> Issues and pull requests are **piling up**, core devs **seem to be absent**

> focused on other projects (**worryingly**, **competing** projects at times)

> **Somehow**, the last 2 releases missed a 5-liner **PR I wrote**

> we are under a lot of pressure to explain continued use of a product that is **poorly supported**

> find more **before everyone has effectively left**

> **deprecate/archive it** so everyone can simply **jump ship**

> **The problem** I see

> **some (currently unresponsive)** maintainer

> **ruining** the community effort

Your messages would have greatly benefitted from having a fair bit less language of entitlement and a fair bit more language of empathy in them.

**Comment by acoulton** [comment]
@everzet thanks for your response. I'd be happy to take on a limited maintainer role as you describe, and of course to work with any others that want to contribute.

Thanks too for offering your time to support onboarding, that would be very welcome. Hopefully I wouldn't need too much from you : although I don't have a lot of contributions here we've been using and following the project for quite a long time (13 or more years I think!) and have built some internal behat extensions/worked on mink drivers etc so I think I have at least a bit of an understanding of how things fit together.

**Comment by uuf6429** [comment]
@everzet I was being factual and to the point.

As I said above, (which unfortunately you left out), I need to explain and defend the choice of using this software. Users', (such as myself), reputation and career depends (in part) on it.

Speaking of empathy, silence and ghosting towards contributors doesn't invite empathy either (ignoring PRs/issues or abruptly stopping communication and not responding to calls for a status update) - and of course that looks worse when maintainers are seen committing to other (competing) projects.

The tone behind the original wasn't from entitlement, but rather from desperation.

**Comment by uuf6429** [comment]
As to how to continue, I don't mind getting involved, but I don't want to be taking decisions entirely by myself... so I think it makes sense to have a team of people that are willing to dedicate some time.

**Comment by carlos-granados** [comment]
@everzet many thanks for getting back to us, it is great to finally be able to hear from one of the owners of the repository. I would like to start by expressing my thanks for creating such a wonderful tool which has helped so many developers. Please excuse us if any of our words sounded harsh, they were mainly driven by desperation, as @uuf6429 said. It was very disheartening to see the months pass and that no one was available to show some love and care for this project.

But let's put that behind and look and how to proceed from here. I think that instead of naming a single maintainer, it would be best to name a group of them. In that way the burden and responsibility is shared and makes it more difficult for the project not to have a responding owner in the future. I think that three would be an ideal number. I volunteer to be one of them.

Once that is decided, I would say that we first need to start by looking at all the pending pull requests and decide which ones can be merged, which need more work and which should be closed. I would say that a couple of rules should apply:
- Any approved or rejected PR needs to have been approved or rejected at least by two of the new maintainers
- The maintainers should not be able to approve or merge their own PRs
After that, work on new PRs should continue as proactively as possible. There is nothing more disheartening than spending some time to create an improvement for a project, only for it to be left unanswered for weeks or months

Then we need to look at current issues and triage them, deciding which ones should be closed, which should be given priority and which need more work, labelling them as needed. Again I would suggest having the opinion of at least two maintainers before acting on any issue

Then we can start working on these issues, starting with the ones with high priority and bugs. There should be a way for any maintainer to set ownership of one issue to indicate that they are going to work on them.

And then, as we gain knowledge and confidence, we can start looking into new features and future improvements. @ciaranmcnulty suggested a number of future improvements in this discussion: https://github.com/Behat/Behat/discussions/1383 He has not been active recently in this repository but I am hoping that if he sees that there is new life in this project he will be willing to participate again as he certainly seems to have a really deep knowledge of it

Finally I think that there should be a channel where the maintainers could communicate and discuss issues, plans, features, etc... I am thinking of maybe a slack channel but open to other suggestions.

What do you all think?

**Comment by acoulton** [comment]
@carlos-granados that sounds a sensible plan to me, the only couple of things I'd say are:

* Requiring 2 approvals makes sense and is easily supported on github. I'm not sure the tooling so easily supports multiple rejection (e.g. tracking the state of what's outstanding, who still needs to look at things), and sometimes an issue / PR will be an obvious "no" and might not be a good use of time to force multiple maintainers to look at it. Closing an issue / PR has no immediate impact on project users, and can easily be undone - so I suggest maintainers should be able to close things individually, giving a reason and making clear that the contributor is welcome to "appeal" the decision at which point another maintainer could reopen if they think appropriate. This will help us keep the list of issues & PRs manageable and from a contributor perspective getting a definite "no, we won't be doing that" is much more helpful than being left in limbo as it allows people to find other solutions.
* Maintainers should not be able to merge their own PRs, but should only require approval from one other maintainer. Otherwise if we only have 3, then a pull by a maintainer requires all 3 maintainers to be involved. That could cause issues if someone is not available, and it seems odd to have a higher bar for contributions from a maintainer than from an external contributor?
* Possibly we should enforce the maintainer-approval rules on an "honour" basis rather than e.g. by locking down the github branch protections? As an open source project, even with 3 maintainers there is a risk we lose maintainers again in the future, and we ideally want to avoid being stuck if that happens. Hopefully of course the remaining maintainer(s) would be able to recruit someone else to join the team, but that might take some time and it would be annoying if the project was again unable to merge even basic / non-controversial dependency or language-version-support updates in that period.

**Comment by carlos-granados** [comment]
@acoulton good points and I agree that we should probably just agree on some "honour" basis for the approval rules as that provides more flexibility

**Comment by carlos-granados** [comment]
Just noticed that Behat has a Gitter channel. I have never used Gitter before but seems similar to Slack in some ways. Could this be the best place to host the communication amongst maintainers? 🤔

**Comment by acoulton** [comment]
I've never used Gitter, willing to give it a go if it suits others (and the fact it is already set up is a bonus)

**Comment by everzet** [comment]
Why not use https://github.com/Behat/Behat/discussions ? Seems much closer. I don't think Gitter experiment ever worked. Might be wrong though.

**Comment by everzet** [comment]
I'll have a look at setting up rulesets and teams sometime this week.

**Comment by carlos-granados** [comment]
> Why not use https://github.com/Behat/Behat/discussions ? Seems much closer. I don't think Gitter experiment ever worked. Might be wrong though.

Ah, did not think of that. May be the right channel, yes

**Comment by AlexSkrypnyk** [comment]
Hi everyone

I use Behat in my Drupal projects daily and I am a maintainer of https://github.com/drevops/behat-steps

I would like to join the maintenance team and also would like to offer some limited monthly sponsorship.

**Comment by carlos-granados** [comment]
Hi @AlexSkrypnyk that's great to hear.

**Comment by everzet** [comment]
@carlos-granados I invited you and @acoulton. Let's stick with just that for a little bit before opening flood gates for more maintainers. Want to make sure our workflows and permissions work and that we have at least 2 good maintainers.

I set the ruleset for Behat and Gherkin repos to require two approvals on PRs and disallow direct pushes. So that should fulfil your policy requirements. I don't believe there is a way to forbid PR initiator to be disallowed to be approval in automatic fashion. So we will have to police this manually.

I also noticed that both Behat and Gherkin pipelines were not in a working state with a bunch of deprecation warnings and version conflicts. I have now resolved all of them so that you can start more or a less from a working state. I cut new Gherkin release, but stopped myself from cutting Behat release, allowing you to merge a bunch more PRs and hold the honours :)

I can't commit much time, but, as promised, I'll be sticking around for a little bit to make sure things are running smoothly before I disengage again.

**Comment by uuf6429** [comment]
It feels that the last few comments answer and satisfy the main points of this issue, so I'm closing it.

Thanks everyone.

**Comment by carlos-granados** [comment]
Thanks @everzet I will try to find some time this week to start working on the repo. @acoulton I'll keep you informed

**Comment by acoulton** [comment]
@everzet accepted the invite, thanks - and thanks for getting the pipelines working for us :)

**Comment by ciaranmcnulty** [comment]
Thanks @everzet , and thanks @acoulton and @carlos-granados for stepping forwards
